### ISSUE TITLE
Windows: Access denied while building project initialised with build-exe

### ISSUE THREAD

**Issue by Einh06** [ziglang/zig#8362]
Tested with`zig version: 0.7.1` and `zig version: 0.8.0-dev.1548+0d96a284e`
I'm currently running into an issue building on Windows where running `zig build` on a project generated with `zig init-exe` fails with `error: unable to build stage1 zig object: AccessDenied`. I ran the command both as user and administrator, to no avail.
The permissions on all folders also seem correct (`%LOCALAPPDATA%\zig` has all permisions, so does my directory with the zig binaries, and the project directory).
`zig build-exe main.zig` also fails, but `zig run main.zig` works.

**Comment by csharpdf** [comment]
I just started learning Zig, and after trying to build a basic app I got this same issue. I tried both the current master branch, `zig-windows-x86_64-0.9.0-dev.1610+e5bc09240`, and the current stable release, `0.8.1`. I'm on Windows 10, and also ran `zig build` on an app initialized with `zig init-exe`. I'd really like to see this issue resolved so I can start working with the language.

**Comment by csharpdf** [comment]
Update: It seems that `zig run src/main.zig` works for the default `zig init-exe` example, but again, the `zig build (either build.zig, src/main.zig, or no param)` still results in an error. This may help a bit @andrewrk

**Comment by aotmr** [comment]
I'm seeing the same issue myself on the stable release, `0.8.1`, both for `zig build run` and `zig run src/main.zig`. This happens both for Powershell and CMD alike from within Visual Studio Code. The issue is intermittent--if I re-try the build, eventually it succeeds.

**Comment by csharpdf** [comment]
It does not work, and even if it did, it's still a major issue. You should be able to build programs on the first attempt with no problems. I sincerely hope that this issue gets fixed. If I could help, I would.

**Comment by aotmr** [comment]
@csharpdf I've been using the Windows builds, extracted to a folder on my drive and added to the PATH. I will see if the problem still exists with the new release.

**Comment by aotmr** [comment]
I'm afraid I get the same issue with 0.9.0, but with a perhaps better error message:
```
error: unable to generate DLL import .lib file: AccessDenied
```

**Comment by marler8997** [comment]
Doesn't reproduce on my machine:
```
C:\temp\b>zig init-exe
info: Created build.zig
info: Created src\main.zig
info: Next, try `zig build --help` or `zig build run`

C:\temp\b>zig build

C:\temp\b>zig version
0.9.0-dev.1815+20e19e75f
```

What directory are you in when you run `zig init-exe` and `zig-build`.  Can you send the commands/output you get?

**Comment by aotmr** [comment]
I extracted the toolchain to `C:\zig\`, which is in PATH. I keep my projects in a subfolder of Documents, in this case `C:\Users\James\Documents\repositories\test`. `zig init-exe` within that folder succeeds, but I get varying results with `zig build`. Here's my output from within a non-elevated Powershell.
```
PS C:\Users\James\Documents\repositories\test> zig init-exe
info: Created build.zig
info: Created src\main.zig
info: Next, try `zig build --help` or `zig build run`
PS C:\Users\James\Documents\repositories\test> zig build
error: unable to generate DLL import .lib file: AccessDenied
PS C:\Users\James\Documents\repositories\test> zig build
error: unable to generate DLL import .lib file: AccessDenied
PS C:\Users\James\Documents\repositories\test> zig build
PS C:\Users\James\Documents\repositories\test> zig build
error: AccessDenied
PS C:\Users\James\Documents\repositories\test> zig build run
error: AccessDenied
PS C:\Users\James\Documents\repositories\test> zig build run
error: AccessDenied
PS C:\Users\James\Documents\repositories\test> zig build run
info: All your codebase are belong to us.
```

**Comment by marler8997** [comment]
Interesting.  The first couple errors you are getting are when Zig compiles windows import libraries and saves them in the global cache.  You can get you're `global_cache_dir` by running `zig env`.  If I was able to reproduce this, I would try building a debug version of Zig to first see if that gives you and stack trace or error information, if it doesn't I would start adding more debug to my compiler to track down exactly what function is failing.  Since I can't reproduce this I'm not sure how to help with this.  But first you should see what your global_cache_dir is and maybe check if there are any issues writing files to it?  Maybe try echo'ing a new file to that directory from the terminal you're running `zig build` in?

**Comment by aotmr** [comment]
It does smell of a race condition, given the seemingly random chance of success. I wish I had any experience with the internals of the Zig compiler; where might I find a set of instructions as to building a debug version?

**Comment by marler8997** [comment]
Here on the wiki: https://github.com/ziglang/zig/wiki/Building-Zig-on-Windows

If you run into issues you can also hit me up or whoever else is able to on Discord or IRC.

**Comment by maxxnino** [comment]
> It does smell of a race condition, given the seemingly random chance of success. I wish I had any experience with the internals of the Zig compiler; where might I find a set of instructions as to building a debug version?

Recently, I got the same error too. Base on your comments. So I try edit some big struct, or function. And it's work

**Comment by marler8997** [comment]
user cr1901 on IRC seems to be able to reproduce this issue sometimes on his machine.  It's also intermittent and goes away after running "zig build" a few times.  He was able to see from "ProcMon" that the "Access Denied" occurs in `NtSetInformationFile`.  We also had the address in `zig.exe` where it was called, using a disassembler I was able to figure out this was occurring in `renameatW` in `os.zig` here:

https://github.com/ziglang/zig/blob/59418d1bf6f82866a1c8f3b35fd815bb7add5129/lib/std/os.zig#L2352

The "ProcMon" event tab also contained some information about what was being renamed:

![Procmon64_q0VgfJRg2v](https://user-images.githubusercontent.com/304904/153136436-b64ec52c-6851-4e00-9a28-b0507c37b012.png)

It appears that the error occurs when Zig tries to rename a directory in "zig-cache\tmp\..." to "zig-cache\o\...".  Windows will give an access denied error when trying to rename a directory for a number of reasons.

https://stackoverflow.com/questions/41365318/access-is-denied-when-renaming-folder

For example, if any running process is using that directory (or a subdirectory of it) as their "current directory", or if any process has an file/subdirectory open inside that directory with certain SHARE flags.  I have also seen cases where it appears that Windows seems to hold on to directories even after they are closed for some period of time, in which case a retry loop may be required here.

**Comment by samhattangady** [comment]
I see this issue happen intermittently as well. For me, deleting the local `zig-cache` directory fixes it.

**Comment by slimsag** [comment]
I can reproduce this pretty consistently with my (unfortunately quite large) project. It seems to happen when one library in my project fails to build, I'll get an error indicating the build failed and then any rebuild after that just a generic access denied. Only way to workaround is to wipe the zig-cache directory entirely.

Is there anything useful I can do to help track this down?

**Comment by marler8997** [comment]
@slimsag I got a little discouraged when I looked into this more.  It seems to me that Windows might have some extremely bad timing issues.  My next step is to take Zig out of the equation and come up with a set of example in `C/C++` that showcase exactly what those issues are.  But I think I need to find a slower machine to do this on because it seems to happen rarely if at all on my very fast machine.  In any case, once we really identify exactly what use cases have the issue, then we can talk about how Zig can address the issue.  The solution may even be filing a bug with Microsoft with the examples we come up with and seeing if they can be fixed.

**Comment by jcmoyer** [comment]
I'm running into the same issue very frequently since a couple months back, but it doesn't seem to occur with any sort of pattern. Sometimes it happens after a compile error, other times it happens even after `zig build run` exits successfully.

**Comment by jcmoyer** [comment]
This sequence of commands consistently reproduces the problem for me:

```
d:\Scratch\access_test>zig init-exe
info: Created build.zig
info: Created src\main.zig
info: Next, try `zig build --help` or `zig build run`

d:\Scratch\access_test>gvim src\main.zig
```

```zig
const std = @import("std");

// force a linker error
extern fn this_symbol_does_not_exist() c_int;

pub fn main() anyerror!void {
    _ = this_symbol_does_not_exist();
    std.log.info("All your codebase are belong to us.", .{});
}

test "basic test" {
    try std.testing.expectEqual(10, 3 + 7);
}
```

```
d:\Scratch\access_test>zig build
lld-link: error: undefined symbol: this_symbol_does_not_exist
>>> referenced by D:\Scratch\access_test\src\main.zig:6
>>>               D:\Scratch\access_test\zig-cache\o\ef6c7e0b8fff0b0784a6055ef2e8b5f8\access_test.obj:(main)
error: LLDReportedFailure
access_test...The following command exited with error code 1:
D:\Scratch\zig-install\bin\zig.exe build-exe D:\Scratch\access_test\src\main.zig --cache-dir D:\Scratch\access_test\zig-cache --global-cache-dir C:\Users\...\AppData\Local\zig --name access_test --enable-cache
error: the following build command failed with exit code 1:
d:\Scratch\access_test\zig-cache\o\0d5803aae53e4d58fb3b4e877aeed5e1\build.exe D:\Scratch\zig-install\bin\zig.exe d:\Scratch\access_test d:\Scratch\access_test\zig-cache C:\Users\...\AppData\Local\zig

d:\Scratch\access_test>zig build
error: AccessDenied
access_test...The following command exited with error code 1:
D:\Scratch\zig-install\bin\zig.exe build-exe D:\Scratch\access_test\src\main.zig --cache-dir D:\Scratch\access_test\zig-cache --global-cache-dir C:\Users\...\AppData\Local\zig --name access_test --enable-cache
error: the following build command failed with exit code 1:
d:\Scratch\access_test\zig-cache\o\0d5803aae53e4d58fb3b4e877aeed5e1\build.exe D:\Scratch\zig-install\bin\zig.exe d:\Scratch\access_test d:\Scratch\access_test\zig-cache C:\Users\...\AppData\Local\zig
```

**Comment by Guigui220D** [comment]
It has happened to me frequently for a while now, it's quite annoying. My workaround is to delete zig-cache when it happens.

**Comment by jagt** [comment]
I'm also running into this recently. I figured in my case it's exactly the reason [stated above](https://github.com/ziglang/zig/issues/8362#issuecomment-1033407110).

However in my case the reason for access denied is easier: it's just the dest dir already exists. And seems on windows `renameatW` doesn't [return `PathAlreadyExists`](https://github.com/ziglang/zig/blob/971ef7b9c2d67ee849e252e75b79ee1e1ef3da6f/lib/std/os.zig#L2692) while [`renameTmpIntoCache` checks only for that](https://github.com/ziglang/zig/blob/971ef7b9c2d67ee849e252e75b79ee1e1ef3da6f/src/link.zig#L726).

I think we should do a check for path already exists in `renameatW`?

**Comment by squeek502** [comment]
Some relevant links:

- Seems like this could potentially be something of a duplicate of https://github.com/ziglang/zig/issues/6364
- I've made a comment with some more potentially relevant details here: https://github.com/ziglang/zig/pull/11413#issuecomment-1095571208

**Comment by slimsag** [comment]
https://github.com/ziglang/zig/commit/b12992cb0172bc2de1f1aae3da57b952dbaf70cd may have fixed this?

**Comment by slimsag** [comment]
https://github.com/ziglang/zig/pull/12526 may help this

**Comment by kubkon** [comment]
Fingers crossed! This actually made `kubkon/zld` run tests on Windows so here goes nothing!

**Comment by sandos** [comment]
Master install zip from 2022-08-23 fixes this issue for me finally!

**Comment by andrewrk** [comment]
Fixed by 4a98385b0aa3808ab05a1ebfbc90fd0bcd97c0d9, thank you @kubkon :pray:

**Comment by marler8997** [comment]
Whoa crazy, how was the `MachO.zig` file being used on Windows to cause this @kubkon ?  Also thanks so much for fixing this!
